src/python/bot/fuzzers/engine.py

```diff
@@ -111,15 +111,17 @@ def reproduce(self, target_path, input_path, arguments, max_time):
     """
     raise NotImplementedError
 
-  def minimize_corpus(self, target_path, arguments, input_dirs, output_dir,
-                      reproducers_dir, max_time):
+  def minimize_corpus(self, target_path, arguments, existing_corpus_dirs,
+                      new_corpus_dir, output_corpus_dir, reproducers_dir,
+                      max_time):
     """Optional (but recommended): run corpus minimization.
 
     Args:
       target_path: Path to the target.
       arguments: Additional arguments needed for corpus minimization.
-      input_dirs: Input corpora.
-      output_dir: Output directory to place minimized corpus.
+      existing_corpus_dirs: Input corpora that existed before the fuzzing run.
+      new_corpus_dir: Input corpus that was generated during the fuzzing run.
```

Comment:
this makes the interface harder to understand and implement, so I think it's the wrong place to change this. 

since this is specific to libFuzzer, we should place this in the implementation for that instead, rather than forcing all engines to adopt this convention.

we could e.g. do this in libFuzzer's engine.py by having our own _minimize_corpus() function there which has this extra argument that the more general interface (minimiz_corpus) piggy backs on.

Reply:
+1. yes this shouldn't be in main engine fuzzer interface. also not in libfuzzer's main minimize_corpus interface, it should be in a helper function _minimize_corpus with extra arg that libfuzzer's main minimize_corpus function can call. i prefer to keep existing name and interfaces, esp the input and output ones.

Reply:
Thank you, guys! The problem is that libFuzzer's specific `_merge_new_units` function is the last place where we know which directory has the new inputs. From this point on, we have three options how to pass that knowledge down to the actual minimization code (i.e. `minimize_corpus`):

1) Pass it as an explicit argument -- this is the current patchset.
2) Pass it as an optional argument -- similar approach, though seemed a bit more confusing to me.
3) Pass it as a certain element of the `input_dirs` (e.g. the first or the last) -- seemed fragile to me, as it'd be hard to enforce the order besides leaving a comment in the code.

Since `minimize_corpus` is an optional interface yet to be implemented for fuzzing engines other than libFuzzer, it seemed best to me to extend it by this. An additional argument to support this exception would be the prediction that if we ever add minimization for other fuzzing engines, we'll want to capture accurate stats from it as well, which means we will likely need to know which inputs are "new".

Please let me know if either 2) or 3) seems better than 1) to you. I think I like 2) as it appears to be more flexible, but that's still an extension of the interface. Option 3) doesn't change the interface but feels fragile.

@jonathanmetzman @mbarbella-chromium PTAL as well please.

Reply:
We can discuss it next week, I don't want to bother you while you're out/travelling :)

Reply:
My suggestion was to have something like the following:

```python
def _minimize_corpus_two_step(self, target_path, arguments, existing_corpus_dirs,
                      new_corpus_dir, output_corpus_dir, reproducers_dir,
                      max_time):
   ...

def minimize_corpus(self): # without the additional arg
  # plain minimize without two step.
```

Then call _minimize_corpus_two_step from _merge_new_units instead.

There will likely be some way to structure the code here such that code is shared as much as possible, maybe with _minimize_corpus_two_step calling minimize_corpus, or by creating another general helper which is used by both.

Would this work?

I imagine we won't need this two step minimization all the time, especially once we do the refactoring to make corpus pruning task use this interface instead, where incremental stats aren't needed.

Note that the intention of `minimize_corpus` was always to be general and used for e.g. corpus pruning -- it just happened that we could reuse it in `fuzz()`. Making this interface specific to fuzzing (and libFuzzer) isn't ideal.

Reply:
> Would this work?

Sounds reasonable, I'll try and see how it looks then. Thanks for the suggestion!

> I imagine we won't need this two step minimization all the time, especially once we do the refactoring to make corpus pruning task use this interface instead, where incremental stats aren't needed.

Even though we don't have incremental stats from the pruning, we still have overall corpus stats -- we should re-use the new stats parsing to get the stats from the final line in the merge log, when we switch to calling `minimize_corpus` instead of `runner.merge`.

I also think we may want incremental stats for the pollination cases (will also need to figure out where to store those stats), but it's outside of the scope of this change. //cc @mbarbella-chromium who might be interested in tracking cross-pollination.

> Note that the intention of `minimize_corpus` was always to be general and used for e.g. corpus pruning -- it just happened that we could reuse it in `fuzz()`.

Yeah, I see... even its name is rather confusing in a context of merging new units.